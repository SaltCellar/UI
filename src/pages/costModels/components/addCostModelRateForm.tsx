import {
  FormGroup,
  FormSelect,
  FormSelectOption,
  InputGroup,
  InputGroupText,
  TextInput,
} from '@patternfly/react-core';
import { DollarSignIcon } from '@patternfly/react-icons';
import { MetricHash } from 'api/metrics';
import { Option } from 'pages/costModels/components/logic/types';
import React from 'react';
import { InjectedTranslateProps } from 'react-i18next';

export const isRateValid = (rate: string) =>
  (!isNaN(Number(rate)) && Number(rate) > 0) || rate === '';
export const canSubmit = (rate: string) =>
  rate === '' || isNaN(Number(rate)) || Number(rate) <= 0;

export const unusedRates = (
  metricsHash: MetricHash,
  tiers: { metric: string; measurement: string }[]
) => {
  const tiersMap = tiers.reduce((acc, curr) => {
    const measureObj =
      acc[curr.metric] === undefined
        ? { [curr.measurement]: true }
        : { ...acc[curr.metric], [curr.measurement]: true };
    return {
      ...acc,
      [curr.metric]: measureObj,
    };
  }, {});
  return Object.keys(metricsHash).reduce((acc, mtr) => {
    const availableMeasurements = Object.keys(metricsHash[mtr])
      .filter(
        msr => tiersMap[mtr] === undefined || tiersMap[mtr][msr] === undefined
      )
      .map(msr => metricsHash[mtr][msr]);
    if (availableMeasurements.length === 0) {
      return acc;
    }
    return {
      ...acc,
      [mtr]: availableMeasurements.reduce((acc_, curr_) => {
        return { ...acc_, [curr_.label_measurement]: true };
      }, {}),
    };
  }, {});
};

interface CategorySelectorProps {
  isInvalid?: boolean;
  label: React.ReactNode;
  id: string;
  value: string;
  onChange: (value: string) => void;
  defaultOption?: Option;
  options: Option[];
  isDisabled?: boolean;
  testId?: string;
  helperText?: React.ReactNode;
}

const CategorySelector: React.SFC<CategorySelectorProps> = ({
  label,
  id,
  isDisabled = false,
  value,
  onChange,
  defaultOption,
  options,
  isInvalid = false,
  testId,
  helperText,
}) => (
  <FormGroup
    data-testid={testId}
    label={label}
    fieldId={id}
    helperText={helperText}
  >
    <FormSelect
      validated={!isInvalid ? 'default' : 'error'}
      isDisabled={isDisabled}
      value={value}
      onChange={onChange}
      aria-label={`form selector ${label}`}
      id={id}
    >
      <FormSelectOption
        isDisabled
        value={defaultOption.value}
        label={defaultOption.label}
      />
      {options.map(opt => (
        <FormSelectOption
          key={`${opt.value}`}
          value={opt.value}
          label={opt.label}
        />
      ))}
    </FormSelect>
  </FormGroup>
);

export interface CostTypeSelectorBaseProps extends InjectedTranslateProps {
  value: string;
  onChange: (value: string) => void;
  costTypes: string[];
}

export const CostTypeSelectorBase: React.SFC<CostTypeSelectorBaseProps> = ({
  t,
  value,
  onChange,
  costTypes,
}) => {
  const options = costTypes.map(costType => ({
    value: costType,
    label: t(`cost_models.${costType}`),
  }));

  const defaultOption = {
    value,
    label: t(`cost_models.${value}`),
  };
  return (
    <CategorySelector
      label={t('cost_models.add_rate_form.cost_type')}
      id={'rate-cost-type-selector'}
      value={value}
      onChange={onChange}
      defaultOption={defaultOption}
      options={options}
      helperText={<a href="">{t('cost_models.learn_more')}</a>}
    />
  );
};

interface SelectorBaseProps extends InjectedTranslateProps {
  options: Option[];
  onChange: (value: string) => void;
  value: string;
  isDisabled?: boolean;
  isInvalid?: boolean;
}

export const MetricSelectorBase: React.SFC<SelectorBaseProps> = ({
  t,
  value,
  onChange,
  isDisabled = false,
  options,
  isInvalid = false,
}) => {
  return (
    <CategorySelector
      testId={'metric-selector'}
      label={t(`cost_models.add_rate_form.metric_select`)}
      id={'metric-selector'}
      value={value}
      onChange={onChange}
      defaultOption={{
        label: t('cost_models.add_rate_form.default_option'),
        value: '',
      }}
      options={options}
      isDisabled={isDisabled}
      isInvalid={isInvalid}
    />
  );
};

const MeasurementSelectorBase: React.SFC<SelectorBaseProps> = ({
  t,
  value,
  onChange,
  isDisabled = false,
  options,
  isInvalid = false,
}) => {
  return (
    <CategorySelector
      testId={'measurement-selector'}
      label={t(`cost_models.add_rate_form.measurement_select`)}
      id={'measurement-selector'}
      value={value}
      onChange={onChange}
      defaultOption={{
        label: t('cost_models.add_rate_form.default_option'),
        value: '',
      }}
      options={options}
      isDisabled={isDisabled}
      isInvalid={isInvalid}
    />
  );
};

interface InputBase extends InjectedTranslateProps {
  onChange: (value: string) => void;
  value: string;
  isInvalid?: boolean;
}

const RateInputBase: React.SFC<InputBase> = ({
  t,
  value,
  onChange,
  isInvalid = false,
}) => {
  return (
    <FormGroup
      label={t('cost_models.add_rate_form.rate_input')}
      fieldId="rate-input"
      helperTextInvalid={t('cost_models.add_rate_form.error_message')}
      validated={!isInvalid ? 'default' : 'error'}
    >
      <InputGroup>
        <InputGroupText>
          <DollarSignIcon />
        </InputGroupText>
        <TextInput
          type="text"
          aria-label={t('cost_models.add_rate_form.rate_input')}
          id="rate-input"
          placeholder="0.00"
          value={value}
          onChange={onChange}
          validated={!isInvalid ? 'default' : 'error'}
        />
      </InputGroup>
    </FormGroup>
  );
};

export interface AddCostModelRateFormProps extends InjectedTranslateProps {
  metric: string;
  setMetric: (value: string) => void;
  measurement: string;
  setMeasurement: (value: string) => void;
  rate: string;
  setRate: (value: string) => void;
  metricOptions: Option[];
  measurementOptions: Option[];
  validRate: boolean;
  enableSubmit?: boolean;
  submit?: () => void;
}

export const SetMetric = ({ t, onChange, value, options }) => {
  return (
    <MetricSelectorBase
      t={t}
      onChange={onChange}
      value={value}
      options={options}
    />
  );
};

export const SetMeasurement = ({
  metricChange,
  metric,
  metricOptions,
  measurementChange,
  measurement,
  measurementOptions,
  t,
}) => {
  return (
    <>
      <MetricSelectorBase
        t={t}
        onChange={metricChange}
        value={metric}
        options={metricOptions}
      />
      <MeasurementSelectorBase
        t={t}
        onChange={measurementChange}
        value={measurement}
        options={measurementOptions}
      />
    </>
  );
};

export const SetRate = ({
  metricChange,
  metric,
  metricOptions,
  measurementChange,
  measurement,
  measurementOptions,
  isMeasurementInvalid,
  rate,
  rateChange,
  isRateInvalid,
  costTypes,
  costType,
  costTypeChange,
  t,
}) => {
  return (
    <>
      <MetricSelectorBase
        t={t}
        onChange={metricChange}
        value={metric}
        options={metricOptions}
      />
      <MeasurementSelectorBase
        t={t}
        onChange={measurementChange}
        value={measurement}
        options={measurementOptions}
        isInvalid={isMeasurementInvalid}
      />
      <RateInputBase
        t={t}
        value={rate}
        onChange={rateChange}
        isInvalid={isRateInvalid}
      />
      <CostTypeSelectorBase
        t={t}
        costTypes={costTypes}
        value={costType}
        onChange={costTypeChange}
      />
    </>
  );
};
