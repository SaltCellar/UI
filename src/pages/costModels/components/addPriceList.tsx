import {
  ActionGroup,
  Button,
  ButtonVariant,
  Stack,
  StackItem,
  Text,
  TextContent,
  TextVariants,
  Title,
} from '@patternfly/react-core';
import { Metric, MetricHash } from 'api/metrics';
import { Form } from 'components/forms/form';
import React from 'react';
import { InjectedTranslateProps, translate } from 'react-i18next';
import { connect, Omit } from 'react-redux';
import { createMapStateToProps } from 'store/common';
import { metricsSelectors } from 'store/metrics';
import { assign, interpret, Machine, State } from 'xstate';
import {
  SetMeasurement,
  SetMetric,
  SetRate,
  unusedRates,
} from './addCostModelRateForm';
import { styles } from './addPriceList.styles';

export interface TierData {
  metric: string;
  measurement: string;
  rate: string;
  costType: string;
  meta: Metric;
}

export type SubmitPayload = TierData;

interface AddRateStates {
  states: {
    setMetric: {};
    setMeasurement: {};
    setRate: {
      states: {
        invalid: {};
        valid: {};
        init: {};
      };
    };
  };
}

type AddRateEventPayload = Partial<TierData>;

type AddRateEvents =
  | { type: 'CHANGE_METRIC'; payload: AddRateEventPayload }
  | { type: 'CHANGE_MEASUREMENT'; payload: AddRateEventPayload }
  | { type: 'CHANGE_RATE'; payload: AddRateEventPayload }
  | { type: 'CHANGE_INFRA_COST'; payload: AddRateEventPayload };

type AddRateContext = Omit<TierData, 'meta'>;

export const addRateMachine = Machine<
  AddRateContext,
  AddRateStates,
  AddRateEvents
>(
  {
    id: 'add-new-rate-machine',
    context: {
      rate: '',
      metric: '',
      measurement: '',
      costType: '',
    },
    initial: 'setMetric',
    states: {
      setMetric: {
        on: {
          CHANGE_METRIC: {
            target: 'setMeasurement',
            actions: ['metric'],
          },
        },
        meta: {
          test: ({ queryAllByLabelText }) => {
            const metricSelectors = queryAllByLabelText(
              'cost_models.add_rate_form.metric_select'
            );
            expect(metricSelectors.length).toBe(1);
            const measurementSelectors = queryAllByLabelText(
              'cost_models.add_rate_form.measurement_select'
            );
            expect(measurementSelectors.length).toBe(0);
            const rateInputs = queryAllByLabelText(
              'cost_models.add_rate_form.rate_inputs'
            );
            expect(rateInputs.length).toBe(0);
          },
        },
      },
      setMeasurement: {
        on: {
          CHANGE_METRIC: {
            target: 'setMeasurement',
            actions: ['metric', 'resetMeasurement'],
          },
          CHANGE_MEASUREMENT: {
            target: 'setRate',
            actions: ['measurement'],
          },
        },
        meta: {
          test: ({ queryAllByLabelText }) => {
            const metricSelectors = queryAllByLabelText(
              'cost_models.add_rate_form.metric_select'
            );
            expect(metricSelectors.length).toBe(1);
            const measurementSelectors = queryAllByLabelText(
              'cost_models.add_rate_form.measurement_select'
            );
            expect(measurementSelectors.length).toBe(1);
            const rateInputs = queryAllByLabelText(
              'cost_models.add_rate_form.rate_inputs'
            );
            expect(rateInputs.length).toBe(0);
          },
        },
      },
      setRate: {
        initial: 'init',
        on: {
          CHANGE_METRIC: [
            {
              target: '.invalid',
              actions: ['metric', 'resetMeasurement'],
            },
          ],
          CHANGE_MEASUREMENT: [
            {
              target: '.valid',
              actions: ['measurement'],
              cond: 'isValid',
            },
            {
              target: '.invalid',
              actions: ['measurement'],
            },
          ],
          CHANGE_INFRA_COST: {
            actions: ['cost_type'],
          },
          CHANGE_RATE: [
            {
              target: '.valid',
              actions: ['rate'],
              cond: 'isValid',
            },
            {
              target: '.invalid',
              actions: ['rate'],
            },
          ],
        },
        meta: {
          test: ({ queryAllByLabelText }) => {
            const metricSelectors = queryAllByLabelText(
              'cost_models.add_rate_form.metric_select'
            );
            expect(metricSelectors.length).toBe(1);
            const measurementSelectors = queryAllByLabelText(
              'cost_models.add_rate_form.measurement_select'
            );
            expect(measurementSelectors.length).toBe(1);
            const rateInputs = queryAllByLabelText(
              'cost_models.add_rate_form.rate_input'
            );
            expect(rateInputs.length).toBe(1);
          },
        },
        states: {
          invalid: {},
          valid: {},
          init: {},
        },
      },
    },
  },
  {
    actions: {
      metric: assign({
        metric: (_ctx, evt) => evt.payload && evt.payload.metric,
      }),
      measurement: assign({
        measurement: (_ctx, evt) => evt.payload && evt.payload.measurement,
        costType: (_ctx, evt) => evt.payload && evt.payload.costType,
      }),
      rate: assign({
        rate: (_ctx, evt) => evt.payload && evt.payload.rate,
      }),
      resetMeasurement: assign({
        measurement: (_ctx, _evt) => '',
        costType: (_ctx, _evt) => 'Supplementary',
      }),
      cost_type: assign({
        costType: (_ctx, evt) => evt.payload && evt.payload.costType,
      }),
    },
    guards: {
      isValid: (ctx, evt) => {
        const rateNumber = Number(ctx.rate);
        switch (evt.type) {
          case 'CHANGE_METRIC':
            return (
              !isNaN(rateNumber) &&
              rateNumber > 0 &&
              ctx.rate !== '' &&
              evt.payload &&
              evt.payload.metric !== ''
            );
          case 'CHANGE_MEASUREMENT':
            return (
              !isNaN(rateNumber) &&
              rateNumber > 0 &&
              ctx.rate !== '' &&
              evt.payload &&
              evt.payload.measurement !== ''
            );
          case 'CHANGE_RATE':
            return (
              evt.payload &&
              !isNaN(Number(evt.payload.rate)) &&
              Number(evt.payload.rate) > 0 &&
              ctx.measurement !== '' &&
              evt.payload &&
              evt.payload.rate !== ''
            );
        }
      },
    },
  }
);

interface AddPriceListBaseProps extends InjectedTranslateProps {
  costTypes: string[];
  metricsHash: MetricHash;
  submitRate: (data: SubmitPayload) => void;
  cancel: () => void;
  items: TierData[];
}

export type CurrentStateMachine = State<
  AddRateContext,
  AddRateEvents,
  AddRateStates
>;

interface AddPriceListBaseState {
  current: CurrentStateMachine;
}

export class AddPriceListBase extends React.Component<
  AddPriceListBaseProps,
  AddPriceListBaseState
> {
  public state = {
    current: addRateMachine.initialState,
  };
  public service = interpret(addRateMachine).onTransition(current =>
    this.setState({ current })
  );

  public componentDidMount() {
    this.service.start();
  }

  public componentWillUnmount() {
    this.service.stop();
  }

  public renderForm() {
    const {
      current: {
        context: { metric, measurement, rate, costType },
      },
    } = this.state;
    const { t, items, metricsHash, costTypes } = this.props;
    const { send } = this.service;
    const stateNames = this.state.current.toStrings();
    const mainState = stateNames.length > 1 ? stateNames[1] : stateNames[0];

    const availableRates = unusedRates(metricsHash, items);

    switch (mainState) {
      case 'setMetric':
        return (
          <SetMetric
            t={t}
            options={Object.keys(availableRates).map(r => ({
              label: t(`cost_models.${r}`),
              value: r,
            }))}
            onChange={(value: string) =>
              send({ type: 'CHANGE_METRIC', payload: { metric: value } })
            }
            value={metric}
          />
        );
      case 'setMeasurement':
        return (
          <SetMeasurement
            t={t}
            metricOptions={Object.keys(availableRates).map(r => ({
              label: t(`cost_models.${r}`),
              value: r,
            }))}
            metricChange={(value: string) =>
              send({ type: 'CHANGE_METRIC', payload: { metric: value } })
            }
            metric={metric}
            measurementOptions={Object.keys(availableRates[metric]).map(m => ({
              label: t(`cost_models.${m}`, {
                units: metricsHash[metric][m].label_measurement_unit,
              }),
              value: m,
            }))}
            measurement={measurement}
            measurementChange={(value: string) =>
              send({
                type: 'CHANGE_MEASUREMENT',
                payload: {
                  measurement: value,
                  costType: metricsHash[metric][value].default_cost_type,
                },
              })
            }
          />
        );
      case 'setRate.init':
      case 'setRate.valid':
        return (
          <SetRate
            t={t}
            metricOptions={Object.keys(availableRates).map(r => ({
              label: t(`cost_models.${r}`),
              value: r,
            }))}
            metricChange={(value: string) =>
              send({ type: 'CHANGE_METRIC', payload: { metric: value } })
            }
            metric={metric}
            measurement={measurement}
            measurementOptions={Object.keys(availableRates[metric]).map(m => ({
              label: t(`cost_models.${m}`, {
                units: metricsHash[metric][m].label_measurement_unit,
              }),
              value: m,
            }))}
            measurementChange={(value: string) =>
              send({
                type: 'CHANGE_MEASUREMENT',
                payload: {
                  measurement: value,
                  costType: metricsHash[metric][value].default_cost_type,
                },
              })
            }
            rate={rate}
            rateChange={(value: string) =>
              send({ type: 'CHANGE_RATE', payload: { rate: value } })
            }
            isRateInvalid={false}
            isMeasurementInvalid={false}
            costTypes={costTypes}
            costType={costType}
            costTypeChange={value =>
              send({ type: 'CHANGE_INFRA_COST', payload: { costType: value } })
            }
          />
        );
      case 'setRate.invalid':
        return (
          <>
            <SetRate
              t={t}
              metricOptions={Object.keys(availableRates).map(r => ({
                label: t(`cost_models.${r}`),
                value: r,
              }))}
              metricChange={(value: string) =>
                send({ type: 'CHANGE_METRIC', payload: { metric: value } })
              }
              metric={metric}
              measurementOptions={Object.keys(availableRates[metric]).map(
                m => ({
                  label: t(`cost_models.${m}`, {
                    units: metricsHash[metric][m].label_measurement_unit,
                  }),
                  value: m,
                })
              )}
              measurement={measurement}
              measurementChange={(value: string) =>
                send({
                  type: 'CHANGE_MEASUREMENT',
                  payload: {
                    measurement: value,
                    costType: metricsHash[metric][value].default_cost_type,
                  },
                })
              }
              rate={rate}
              rateChange={(value: string) =>
                send({ type: 'CHANGE_RATE', payload: { rate: value } })
              }
              isRateInvalid={
                isNaN(Number(rate)) || rate === '' || Number(rate) <= 0
              }
              isMeasurementInvalid={measurement === ''}
              costTypes={costTypes}
              costType={costType}
              costTypeChange={value =>
                send({
                  type: 'CHANGE_INFRA_COST',
                  payload: { costType: value },
                })
              }
            />
          </>
        );
    }
  }

  public renderActions() {
    const { t, metricsHash, submitRate, cancel } = this.props;
    const {
      current,
      current: {
        context: { metric, measurement, rate, costType },
      },
    } = this.state;

    if (current.matches('setRate.valid')) {
      return (
        <ActionGroup>
          <Button
            data-testid="add-rate-enable"
            variant={ButtonVariant.primary}
            onClick={() =>
              submitRate({
                metric,
                measurement,
                rate,
                costType,
                meta: metricsHash[metric][measurement],
              })
            }
          >
            {t('cost_models_wizard.price_list.add_rate')}
          </Button>
          <Button variant={ButtonVariant.link} onClick={cancel}>
            {t('cost_models_wizard.price_list.cancel')}
          </Button>
        </ActionGroup>
      );
    }
    return (
      <ActionGroup>
        <Button
          data-testid="add-rate-disabled"
          variant={ButtonVariant.primary}
          isDisabled
        >
          {t('cost_models_wizard.price_list.add_rate')}
        </Button>
        <Button variant={ButtonVariant.link} onClick={cancel}>
          {t('cost_models_wizard.price_list.cancel')}
        </Button>
      </ActionGroup>
    );
  }

  public render() {
    const { t } = this.props;

    return (
      <Stack hasGutter>
        <StackItem>
          <Title headingLevel="h2" size="xl">
            {t('cost_models_wizard.price_list.title')}
          </Title>
        </StackItem>
        <StackItem>
          <TextContent>
            <Text component={TextVariants.h6}>
              {t('cost_models_wizard.price_list.sub_title_add')}
            </Text>
          </TextContent>
        </StackItem>
        <StackItem>
          <Form style={styles.form}>{this.renderForm()}</Form>
        </StackItem>
        <StackItem>{this.renderActions()}</StackItem>
      </Stack>
    );
  }
}

export default connect(
  createMapStateToProps(state => ({
    costTypes: metricsSelectors.costTypes(state),
  }))
)(translate()(AddPriceListBase));
