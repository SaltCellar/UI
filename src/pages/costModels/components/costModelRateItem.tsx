import {
  DataListAction,
  DataListCell,
  DataListItem,
  DataListItemCells,
  DataListItemRow,
  Title,
} from '@patternfly/react-core';
import React from 'react';
import { InjectedTranslateProps, translate } from 'react-i18next';
import { formatCurrency } from 'utils/rateCurrency';

export interface CostModelRateItemProps extends InjectedTranslateProps {
  index: number;
  units: string;
  metric: string;
  measurement: string;
  rate: string;
  actionComponent?: React.ReactNode;
}

const CostModelRateItemBase: React.SFC<CostModelRateItemProps> = ({
  t,
  index,
  units,
  metric,
  measurement,
  rate,
  actionComponent,
}) => {
  const unitsLabel = t(`cost_models.${units}`);
  return (
    <DataListItem aria-labelledby={`rate-${index}`}>
      <DataListItemRow>
        <DataListItemCells
          dataListCells={[
            <DataListCell key={`rate-data`}>
              <Title headingLevel="h2" size="lg">
                {t(`cost_models.${metric}`)}{' '}
                {t(`cost_models.lowercase.${measurement}`, {
                  units: unitsLabel,
                })}
              </Title>
              <Title headingLevel="h2" size="md">
                {t(`cost_models.for_every`, {
                  units: unitsLabel,
                  rate: formatCurrency(Number(rate)),
                })}
              </Title>
            </DataListCell>,
          ]}
        />
        {Boolean(actionComponent) && (
          <DataListAction
            aria-label={`actions rate-${index}`}
            id={`rate-actions-${index}`}
            aria-labelledby={`rate-${index}`}
          >
            {actionComponent}
          </DataListAction>
        )}
      </DataListItemRow>
    </DataListItem>
  );
};

export default translate()(CostModelRateItemBase);
