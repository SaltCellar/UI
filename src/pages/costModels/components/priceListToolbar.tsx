import {
  Toolbar,
  ToolbarContent,
  ToolbarFilter,
  ToolbarGroup,
  ToolbarItem,
  ToolbarItemVariant,
} from '@patternfly/react-core';
import React from 'react';

interface PriceListToolbarProps {
  primary: React.ReactNode;
  selected: string;
  secondaries: {
    component: React.ReactNode;
    name: string;
    onRemove: (category: string, chip: string) => void;
    filters: string[];
  }[];
  onClear: () => void;
  pagination: React.ReactNode;
  button: React.ReactNode;
}

export const PriceListToolbar: React.SFC<PriceListToolbarProps> = ({
  primary,
  secondaries,
  pagination,
  button,
  onClear,
  selected,
}) => {
  return (
    <Toolbar
      clearAllFilters={onClear}
      id="price-list-toolbar"
      style={{ marginBottom: '10px', marginTop: '10px' }}
    >
      <ToolbarContent>
        <ToolbarGroup variant="filter-group">
          <ToolbarItem>{primary}</ToolbarItem>
          {secondaries.map(secondary => {
            return (
              <ToolbarItem key={secondary.name}>
                <ToolbarFilter
                  deleteChip={secondary.onRemove}
                  chips={secondary.filters}
                  categoryName={secondary.name}
                >
                  {selected === secondary.name ? secondary.component : ''}
                </ToolbarFilter>
              </ToolbarItem>
            );
          })}
        </ToolbarGroup>
        <ToolbarItem>{button}</ToolbarItem>
        <ToolbarItem variant={ToolbarItemVariant.pagination}>
          {pagination}
        </ToolbarItem>
      </ToolbarContent>
      <hr className="pf-c-divider" />
    </Toolbar>
  );
};
