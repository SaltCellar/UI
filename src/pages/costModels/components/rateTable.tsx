import { compoundExpand, IActions, Table, TableBody, TableHeader, TableVariant } from '@patternfly/react-table';
import { Rate } from 'api/rates';
import { WithT } from 'i18next';
import React from 'react';
import { formatCurrency } from 'utils/formatValue';

import TagRateTable from './tagRateTable';

interface RateTableProps extends WithT {
  tiers: Rate[];
  actions?: IActions;
  isCompact?: boolean;
}

export const RateTable: React.SFC<RateTableProps> = ({ t, tiers, actions, isCompact }) => {
  const [expanded, setExpanded] = React.useState({});
  const cells = [
    { title: t('cost_models.table.description') },
    { title: t('cost_models.table.metric') },
    { title: t('cost_models.table.measurement') },
    { title: t('cost_models.table.calculation_type') },
    { title: t('cost_models.table.rate'), cellTransforms: [compoundExpand] },
  ];
  let counter = 0;
  const rows = tiers.reduce((acc, tier, ix) => {
    const rateKind = tier.tiered_rates ? 'regular' : 'tagging';
    let compoundRows = [];
    if (rateKind === 'tagging') {
      compoundRows = [
        {
          compoundParent: 4,
          parent: ix + counter,
          cells: [
            {
              title: <TagRateTable tagRates={tier.tag_rates} />,
              props: { colSpan: 6, className: 'pf-m-no-padding' },
            },
          ],
        },
      ];
      counter += 1;
    }
    const isOpen = rateKind === 'tagging' ? expanded[ix + counter - 1] || false : undefined;
    return [
      ...acc,
      {
        data: { index: ix, hasChildren: rateKind === 'tagging' },
        cells: [
          tier.description || '',
          tier.metric.label_metric,
          t(`cost_models.${tier.metric.label_measurement}`, {
            units: tier.metric.label_measurement_unit,
          }),
          tier.cost_type,
          {
            title:
              rateKind === 'regular'
                ? `${formatCurrency(Number(tier.tiered_rates[0].value), 'USD')}`
                : t('cost_models.table.tagged_rates'),
            props: { isOpen, style: { padding: rateKind === 'tagging' ? '' : '1.5rem 1rem' } },
          },
        ],
      },
      ...compoundRows,
    ];
  }, []);
  const onExpand = (_event: any, rowIndex: number, _colIndex: number, isOpen: boolean) => {
    setExpanded({ ...expanded, [rowIndex]: !isOpen });
  };
  const actionResolver = rowData => {
    if (rowData.compoundParent) {
      return [];
    }
    return actions;
  };
  return (
    <Table
      aria-label="price list"
      variant={isCompact ? TableVariant.compact : undefined}
      rows={rows}
      cells={cells}
      actionResolver={actionResolver}
      onExpand={onExpand}
    >
      <TableHeader />
      <TableBody />
    </Table>
  );
};
