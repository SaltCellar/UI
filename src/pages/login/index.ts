import { hot } from 'react-hot-loader';
import Login from './login';

export default hot(module)(Login);
