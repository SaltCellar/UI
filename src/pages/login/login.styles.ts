import { StyleSheet } from '@patternfly/react-styles';
import {
  global_BackgroundColor_dark_100,
  global_spacer_xxl,
} from '@patternfly/react-tokens';

export const styles = StyleSheet.create({
  body: {
    backgroundColor: global_BackgroundColor_dark_100.value,
    minHeight: '100vh',
  },
  loginPage: {
    height: '100vh',
  },
  loginBox: {
    padding: global_spacer_xxl.value,
  },
});
