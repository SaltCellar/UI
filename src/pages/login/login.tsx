import {
  Bullseye,
  Button,
  ButtonType,
  ButtonVariant,
  Title,
  TitleSize,
} from '@patternfly/react-core';
import { css } from '@patternfly/react-styles';
import { Box, BoxBody, BoxHeader } from 'components/box';
import { FormGroup } from 'components/formGroup';
import { TextInput } from 'components/textInput';
import React from 'react';
import { Helmet } from 'react-helmet';
import { InjectedTranslateProps, translate } from 'react-i18next';
import { connect } from 'react-redux';
import { createMapStateToProps, FetchStatus } from 'store/common';
import { sessionActions, sessionSelectors } from 'store/session';
import { getTestProps, testIds } from 'testIds';
import { styles } from './login.styles';

export interface Props extends InjectedTranslateProps {
  fetchStatus: FetchStatus;
  login: typeof sessionActions.login;
}

interface State {
  username: string;
  password: string;
}

export class Login extends React.Component<Props, State> {
  public state: State = {
    username: '',
    password: '',
  };

  private handleUsernameChange = (username: string) => {
    this.setState(() => ({
      username,
    }));
  };

  private handlePasswordChange = (password: string) => {
    this.setState(() => ({
      password,
    }));
  };

  private handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    this.props.login({
      username: this.state.username,
      password: this.state.password,
    });
  };

  public render() {
    const { t } = this.props;
    const { username, password } = this.state;
    return (
      <>
        <Helmet>
          <body className={css(styles.body)} />
        </Helmet>
        <div className={css(styles.loginPage)}>
          <Bullseye>
            <Box className={css(styles.loginBox)}>
              <BoxHeader>
                <Title size={TitleSize['2xl']}>{t('login.title')}</Title>
              </BoxHeader>
              <BoxBody>
                <form
                  {...getTestProps(testIds.login.form)}
                  onSubmit={this.handleSubmit}
                >
                  <FormGroup label={t('login.username_label')}>
                    <TextInput
                      {...getTestProps(testIds.login.username_input)}
                      autoFocus
                      isFlat
                      type="text"
                      onChange={this.handleUsernameChange}
                      value={username}
                    />
                  </FormGroup>
                  <FormGroup label={t('login.password_label')}>
                    <TextInput
                      {...getTestProps(testIds.login.password_input)}
                      isFlat
                      type="password"
                      onChange={this.handlePasswordChange}
                      value={password}
                    />
                  </FormGroup>
                  <div>
                    <Button
                      {...getTestProps(testIds.login.submit)}
                      type={ButtonType.submit}
                      variant={ButtonVariant.primary}
                    >
                      {t('login.submit')}
                    </Button>
                  </div>
                </form>
              </BoxBody>
            </Box>
          </Bullseye>
        </div>
      </>
    );
  }
}

export default connect(
  createMapStateToProps(state => ({
    fetchStatus: sessionSelectors.selectLoginFetchStatus(state),
  })),
  {
    login: sessionActions.login,
  }
)(translate()(Login));
