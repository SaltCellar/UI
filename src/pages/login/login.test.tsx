import { AxiosError, AxiosResponse } from 'axios';
import { TextInput } from 'components/textInput';
import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';
import { FetchStatus } from 'store/common';
import { Login, Props } from './login';

const props: Props = {
  fetchStatus: FetchStatus.none,
  login: jest.fn(),
  t: jest.fn(v => v),
};

test('updates username on input change', () => {
  const view = shallow(<Login {...props} />);
  const input = 'Stephen Falken';
  getUsernameInput(view).simulate('change', input);
  expect(getUsernameInput(view).props().value).toBe(input);
});

test('updates password on input change', () => {
  const view = shallow(<Login {...props} />);
  const input = 'Joshua';
  getPasswordInput(view).simulate('change', input);
  expect(getPasswordInput(view).props().value).toBe(input);
});

test('triggers login on form submit', () => {
  const view = shallow(<Login {...props} />);
  const username = 'Stephen Falken';
  const password = 'Joshua';
  const preventDefault = jest.fn();
  getUsernameInput(view).simulate('change', username);
  getPasswordInput(view).simulate('change', password);
  view.find('form').simulate('submit', { preventDefault });
  expect(preventDefault).toBeCalled();
  expect(props.login).toBeCalledWith({
    username,
    password,
  });
});

test('triggers login error on form submit', () => {
  const mockAxiosResponse: AxiosResponse = {
    data: {
      non_field_errors: 'Unable to log in with provided credentials.',
    },
    status: 1,
    statusText: '',
    headers: {},
    config: {},
  };
  const mockAxiosError: AxiosError = {
    config: {},
    message: '',
    name: '',
    response: mockAxiosResponse,
  };
  const view = shallow(<Login {...props} error={mockAxiosError} />);
  const preventDefault = jest.fn();
  view.find('form').simulate('submit', { preventDefault });
  expect(props.login).toBeCalled();
  expect(view.find('.pf-c-alert')).not.toBeNull();
});

function getUsernameInput(view: ShallowWrapper<any, any>) {
  return view.find(TextInput).at(0);
}

function getPasswordInput(view: ShallowWrapper<any, any>) {
  return view.find(TextInput).at(1);
}
