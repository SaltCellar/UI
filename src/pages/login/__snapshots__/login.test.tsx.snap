// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert is shown with login error 1`] = `
<Alert
  action={null}
  className=""
  data-testid="alert"
  title=""
  variant="danger"
>
  Unable to log in with provided credentials.
</Alert>
`;
